dpe/src/x509.rs

```diff
@@ -1135,6 +1408,118 @@ impl X509CertWriter<'_> {
 
         Ok(bytes_written)
     }
+
+    /// Encode a certification request info
+    ///
+    /// Returns number of bytes written to `scratch`
+    ///
+    /// CertificationRequestInfo ::= SEQUENCE {
+    /// version       INTEGER { v1(0) } (v1,...),
+    /// subject       Name,
+    /// subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
+    /// attributes    [0] Attributes{{ CRIAttributes }}} }
+    ///
+    /// # Arguments
+    ///
+    /// * `pubkey` - ECDSA Public key.
+    /// * `subject_name` - The subject name RDN struct to encode.
+    /// * `measurements` - DPE measurement data.
+    pub fn encode_certification_request_info(
+        &mut self,
+        pub_key: &EcdsaPub,
+        subject_name: &Name,
+        measurements: &MeasurementData,
+    ) -> Result<usize, DpeErrorCode> {
+        let cert_req_info_size = Self::get_certification_request_info_size(
+            subject_name,
+            pub_key,
+            measurements,
+            /*tagged=*/ false,
+        )?;
+
+        // CertificationRequestInfo Sequence
+        let mut bytes_written = self.encode_tag_field(Self::SEQUENCE_TAG)?;
+        bytes_written += self.encode_size_field(cert_req_info_size)?;
+
+        // version
+        bytes_written += self.encode_version()?;
+
+        // subject
+        bytes_written += self.encode_rdn(subject_name)?;
+
+        // subjectPublicKeyInfo
+        bytes_written += self.encode_ecdsa_subject_pubkey_info(pub_key)?;
+
+        // extensions
+        bytes_written += self.encode_extensions(measurements)?;
+
+        Ok(bytes_written)
+    }
+
+    /// Encode an PKCS #10 CSR
+    ///
+    /// Returns number of bytes written to `scratch`
+    ///
+    /// CertificateRequest  ::=  SEQUENCE  {
+    ///    certificationRequestInfo       CertificationRequestInfo,
+    ///    signatureAlgorithm             AlgorithmIdentifier,
+    ///    signatureValue                 BIT STRING  }
+    pub fn encode_csr(
+        &mut self,
+        cert_req_info: &[u8],
+        sig: &EcdsaSig,
+    ) -> Result<usize, DpeErrorCode> {
+        let csr_size = cert_req_info.len()
+            + Self::get_ecdsa_sig_alg_id_size(/*tagged=*/ true)?
+            + Self::get_ecdsa_signature_size(sig, /*tagged=*/ true)?;
+
+        // CertificateRequest sequence
+        let mut bytes_written = self.encode_tag_field(Self::SEQUENCE_TAG)?;
+        bytes_written += self.encode_size_field(csr_size)?;
+
+        // CertificationRequestInfo
+        bytes_written += self.encode_bytes(cert_req_info)?;
+
+        // Alg ID
+        bytes_written += self.encode_ecdsa_sig_alg_id()?;
+
+        // Signature
+        bytes_written += self.encode_ecdsa_signature(sig)?;
+
+        Ok(bytes_written)
+    }
+
+    /// Encode a CMS ContentInfo message
+    ///
+    /// ContentInfo ::= SEQUENCE {
+    /// contentType ContentType,
+    /// content [0] EXPLICIT ANY DEFINED BY contentType }
+    pub fn encode_cms(
```

Comment:
Can you add some unit tests to make sure the certs can be parsed?

The certification_request module from x509_parser should work for the PKCS#10 stuff: https://docs.rs/x509-parser/latest/x509_parser/certification_request/index.html

And then the CMS crate you added for parsing the outer CMS part.

Reply:
Done